### PR TITLE
Update TESTING.md

### PR DESCRIPTION
**Pull request by aramaraju**
Added Guardrails test conditions for JB client

## Test plan

Documentation change -reading through and review.

### CODE REVIEW

File: TESTING.md [hunk 1]
```diff
@@ -522,3 +522,8 @@ Repeat the above starting from different themes.
 #### Expected behaviour
 
 Changing theme should lead to full repaint of the colours according to the current theme.
+
+### Guardrails:
+
+- [ ] Check if Guardrails check passes for any code longer than 10 lines generated in Cody Chat
```
Comment (pkukielka) [inline TESTING.md]: Hey, thank for adding test scenario for Guardrails but it looks a bit incomplete to me.
It does not explain what `Guardrails check` is and how to check if it passed.
Ideally it should be in the form:
1. Do <describe exact steps> to trigger some action
2. Check if results looks like that: <describe expected result>

File: TESTING.md [hunk 2]
```diff
@@ -522,3 +522,8 @@ Repeat the above starting from different themes.
...
+- [ ] Check if Guardrails check failes for any code longer than 10 lines that matches indexed OSS
```
Comment (pkukielka) [inline TESTING.md]: `that matches indexed OSS` - what is indexed OSS?
Please let's make testing guide self-eplainatory if possible.

**Comment by aramaraju** [comment]
@pkukielka Does this loom help explain the feature and I can accommodate that into the test plans:

https://www.loom.com/share/077cf0f2b6174c5dbd2825653f6ed4cd?sid=6f76b12b-afd6-4181-802d-2eaa0c35b005

**Comment by pkukielka** [comment]
> @pkukielka Does this loom help explain the feature and I can accommodate that into the test plans:
> 
> https://www.loom.com/share/077cf0f2b6174c5dbd2825653f6ed4cd?sid=6f76b12b-afd6-4181-802d-2eaa0c35b005

Unfortunately I do not have access to this video.
But I'm sure any context added will be helpful.

**Comment by aramaraju** [comment]
@pkukielka sharing the video through [GDrive](https://drive.google.com/file/d/1rtUuAlUvblrnFf5saYlt_pZz3QxuPuCN/view?usp=drivesdk)

**Comment by pkukielka** [comment]
> @pkukielka Does this loom help explain the feature and I can accommodate that into the test plans:
> 
> https://www.loom.com/share/077cf0f2b6174c5dbd2825653f6ed4cd?sid=6f76b12b-afd6-4181-802d-2eaa0c35b005

Thanks, that looks fine.
I found the easiest way to trigger the check is to simply ask chat this question:

```Give me snipped to the javascript opensource code of your choice.```

For me it always returns a bit of code which fails the guardrails.

**Comment by taylorsperry** [comment]
@aramaraju are you still working on updates to this PR and https://github.com/sourcegraph/cody/pull/3204?

**Comment by aramaraju** [comment]
@taylorsperry looks like @pkukielka was able to find a test to verify this. I'm good with that to check for testing for the beta version of it

**Comment by taylorsperry** [comment]
@aramaraju I think it would be good to have this as part of our `TESTING.md` doc too. Automated tests are obviously great, but I want to make sure we have QA eyes on all of our features too. _And_ this is helpful resource for cross-functional teams (docs, marketing, etc) to understand the features the clients support.

**Comment by dominiccooney** [comment]
+1 to @pkukielka 's comments about making the test plan literal with specific actions and expected outputs. It's not going to scale to have a 4 minute Loom for every feature.

**Review by dominiccooney — DISMISSED** [review]
Putting this back in @aramaraju 's queue for @pkukielka 's feedback.

**Comment by aramaraju** [comment]
@cbart Can you suggest a few test cases that you used with the JBs team? You know better the OSS code that's indexed. Thank you :)

**Comment by pkukielka** [comment]
@aramaraju @cbart this is hanging unmerged for 2 months now.
Could you please reword the tests steps (I already give one working example earlier) so we can free you from this one? :)

**Comment by cbart** [comment]
Apologies for not responding more promptly here. I did not see the notifications. I'll work through this today.

File: TESTING.md (hunk 1 above)
Comment (cbart) [reply TESTING.md]: I tried to put in a detailed description of actions, and tangible expectation. Let me know if that works OK.

File: TESTING.md (hunk 2 above)
Comment (cbart) [reply TESTING.md]: For now I dropped this part as I think due to editor changes ([link](https://sourcegraph.slack.com/archives/C05AGQYD528/p1713260328388809)) this is now very hard to test.

**Comment by cbart** [comment]
Done now. The test only goes through success scenario (guardrails check passes).

Testing a guardrails-failed scenario for a generated snippet proves to be extremely hard in the regular use case. Even when giving Cody exact instructions to generate - really rewrite a given snippet it always seems to adjust indentation or make some tiny changes here and there - and attribution check is an exact match.

The way we've been doing this so far was to actually put a snippet in Cody chat ourselves (guardrails runs there too!) and then we have full control over how the snippet looks like, but in both editors that's not possible right now ([link](https://sourcegraph.slack.com/archives/C05AGQYD528/p1713260328388809)).

**Review by pkukielka — APPROVED** [review]
LGTM

**Comment by taylorsperry** [comment]
@aramaraju should we close this out?

**Comment by aramaraju** [comment]
Looks like we capture the tests here from Cezary's changes, I'm okay to close it or merge it

**Comment by taylorsperry** [comment]
Cool, @aramaraju I'm going to approve so I stop getting the notifications, but defer to you on next steps :)

**Review by taylorsperry — APPROVED** [review]

**Comment by Zwingz2019** [comment]
@CodiumAI-Agent /review

**Comment by Zwingz2019** [comment]
@CodiumAI-Agent /describe

**Pull request merged**